File: Template/RMQ/SparseTable.py
"""
SparseTable 模板

可以用來解決可重複貢獻的區間查詢問題，例如區間最大值、區間最小值、區間按位或、區間按位與、區間GCD等。
注意對於區間和這種不可重複貢獻的問題，不能使用SparseTable。
"""

import math

class SparseTable:
    def __init__(self, data, merge_func):
        n = len(data)
        self.data = data
        self.merge = merge_func
        self.max_log = math.floor(math.log2(n)) if n > 0 else 0
        self.st = [[None] * (self.max_log + 1) for _ in range(n)]
        for i in range(n):
            self.st[i][0] = data[i]
        
        for j in range(1, self.max_log + 1):
            for i in range(n):
                self.st[i][j] = self.merge(self.st[i][j - 1], self.st[min(n - 1, i + (1 << (j - 1)))][j - 1])
    
    def query(self, L, R):  # 0-indexed
        k = math.floor(math.log2(R - L + 1))
        return self.merge(self.st[L][k], self.st[R - (1 << k) + 1][k])
    
# Example of usage
A = list(range(1, 11))
st_mx = SparseTable(A, max)
print(st_mx.query(0, 9))
print(st_mx.query(0, 4))
print(st_mx.query(5, 9))

File: Template/RMQ/SparseTable.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

template<typename T, typename Op = function<T(T, T)>>
class SparseTable {
private:
    vector<vector<T>> st;
    Op merge;
    int n;
public:
    SparseTable(const vector<T>& data, Op op) : merge(op), n(data.size()) {
        if (n == 0) return;
        int max_log = __lg(n); 
        st.assign(n, vector<T>(max_log + 1));

        for (int i = 0; i < n; ++i)
            st[i][0] = data[i];

        for (int j = 1; j <= max_log; ++j)
            for (int i = 0; i + (1 << j) <= n; ++i)
                st[i][j] = merge(st[i][j - 1], st[i + (1 << (j - 1))][j - 1]);
    }

    T query(int L, int R) {  // 0-indexed
        int k = __lg(R - L + 1);
        return merge(st[L][k], st[R - (1 << k) + 1][k]);
    }
};

int main() {
    // Example of usage
    vector<int> A = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    auto func = [](int a, int b) { return max(a, b); };
    SparseTable<int, decltype(func)> st(A, func);
    cout << "Max of range [0, 9]: " << st.query(0, 9) << endl;
    cout << "Max of range [0, 4]: " << st.query(0, 4) << endl;
    cout << "Max of range [5, 9]: " << st.query(5, 9) << endl;
    return 0;
}